addons/sale_stock_margin/models/sale_order_line.py

```diff
@@ -9,18 +9,19 @@ class SaleOrderLine(models.Model):
 
     @api.depends('move_ids', 'move_ids.stock_valuation_layer_ids', 'move_ids.picking_id.state')
     def _compute_purchase_price(self):
-        lines_without_moves = self.browse()
-        for line in self:
+        lines_without_moves = self.filtered(lambda l: not l.has_valued_move_ids())
+        lines_with_std_cost = self.filtered(
+            lambda l:
+            l.product_id.with_company(l.company_id) and
+            l.product_id.categ_id.property_cost_method == 'standard'
+        )
+        for line in (self - lines_without_moves - lines_with_std_cost):
             product = line.product_id.with_company(line.company_id)
-            if not line.has_valued_move_ids():
-                lines_without_moves |= line
-            elif product and product.categ_id.property_cost_method != 'standard':
-                purch_price = product._compute_average_price(0, line.product_uom_qty, line.move_ids)
-                if line.product_uom and line.product_uom != product.uom_id:
-                    purch_price = product.uom_id._compute_price(purch_price, line.product_uom)
-                to_cur = line.currency_id or line.order_id.currency_id
-                line.purchase_price = line._convert_to_sol_currency(
-                    purch_price,
-                    product.cost_currency_id,
-                )
-        return super(SaleOrderLine, lines_without_moves)._compute_purchase_price()
+            purch_price = product._compute_average_price(0, line.product_uom_qty, line.move_ids)
+            if line.product_uom and line.product_uom != product.uom_id:
```

Comment:
```suggestion
            if line.product_uom != product.uom_id:
```
There's a SQL constraint on `product_uom` being `NOT NULL`[^1] for any line that gets handled here (those with a `product_id`), so no need to test for its existence.

[^1]: https://github.com/odoo/odoo/blob/dc2821d768a7b4afbb5918e4343c00abeeb31670/addons/sale/models/sale_order_line.py#L24